[WIP] Support custom fields

nrbnlulu opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description
Disclaimer:  _This PR was started as a fix for #920 and expanded to refactoring of the strawberry type system._

### Motivation:
While walking threw the code I have found that being able to extend `StrawberryField` was pretty hard for these reasons:
 1. the field evaluation process relied mostly on functions (namely `_get_fields`).
 2. `StrawberryField` was inheriting `dataclasses.Field` and that IMO unnecessary burden, results namespace clashes, tights strawberry to dataclasses too much and finally it is not a dataclass like many other parts of strawberry.
 3. generic types was evaluated in a very complex way (`copy_with`) and it is not clear at a first glance when a type is already evaluated or not,
 4.  Integrations (like pydantic) are building lots of things on their own and some functionalities that strawberry provides or not allows might not be covered in the integration, as well as they have lots of redundant code that mimics what strawberry-core already does. 
 5. Usage of `@property` was quite extensive and that is a runtime burden and makes hard to deduce what will a value be at a given point.   
 6.  there are probably more things I forgot but the general idea was to padronize the way strawberry does things and rely more on classes than on functions to allow extending.
___
### Implementation details:
#### Generics:
- if a class inherits `Generic` it would be created as `class TemplateTypeDefinition(TypeDefinition)` This class can generate a whole new "`StrawberryObject`" based on `__args__`. the generator is `TemplateTypeDefinition.generate(<types>)`. it is lunched by the class `__getitem__` that defined in `StrawberryMeta`, so  i.e `Apple[int]` is equivalent to `Apple._type_definition.generate((int, ))`.
- I introduced new `validate` abstract method to `StrawberryType` it is used to determine generic unions.   
  *NOTE*:  With the new API for generating templates this might not even be necessary because users would be able to do this:
 ```py
@strawberry.type
class A(Generic[T]):
    a: T

@strawberry.type
class SomeType():
    @strawberry.field
    def union_generic(self) -> Union[A[int], A[str]]
        return A[int](1)
 ```
 though this would break existing code.
- generics are now cached by `__args__` signature and every type is created just once. This is tested using `timeit`. 

#### Scalars:
 I removed `ScallarWrapper` as it made issued for me with the new `validate` method of `StrawberryType`, every scalar now is just a `ScalarDefinition`, I would consider renaming this to `StrawberryScalar`.
 
#### strawberry.ID
`strawberry.ID` is no longer a plain `NewType` but a `ScalarDefinition` using implementation from graphql-core, I saw that is is a 'TODO' and also it did issues with `validate()` of generic unions.  

#### pydantic:
I have also refactored pydantic integration, and it is no longer building things on it's own, it connects to hooks from strawberry-core to append and modify fields.

#### StrawberryObject:
strawberry object is what will lay behind `@strawberry.type` and it is extending the `StrawberryMeta` class to evaluate generics as explained above.
This has the benifit of duck-typing `_type_definition` and allows integrations to connect to strawberry type creation system.
The workflow is like this:
1. `StrawberryObject` creates a new type from the decorated class that extends `StrawberryObject`.
2. `StrawberryObject` calls a customizable function to create the `TypeDefinition` (to support extending by integrations). 
3. `TypeDefinition` will find all the fields in the class including pre-evaluated dataclasses fields.
7. `TypeDefinition` will create a `StrawberryField` for each field (integrations can define their own `StrawberryField` extended and it will be created instead).
8. `TypeDefinition` will call `__pre_dataclass_creation__` to allow integrations to add  or modify fields.
9. dataclass is created and the rest is history.

As you can see this PR won't only allow custom fields but also custom `StrawberryObject` and `TypeDefinition` the pydantic integration can set an example for this.
___

Please contact me if You want more information, Thank you @patrick91, @BryceBeagle  for your valuable time and patience.

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your pull request introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

* #920 

## Checklist


<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### tests
- [x] renamer field
- [x] type changer field
- [x] argument modification field
- [x] override `get_result` field
- [ ] batched fields.
- [ ] field "`__post_init__`" 
- [x] not supported resolver assertion
- [x] not supported evolvable 
- [ ] StrawberryField was initialized with origin but with no python name assertion.
- [ ] test validate generic union with `StrawberryObject`

botberry commented:
Thanks for adding the `RELEASE.md` file!

![](https://media.giphy.com/media/xq1FxHkABwW7m/giphy.gif)

Here's a preview of the changelog:

---

This release will try to address [#1100](https://github.com/strawberry-graphql/strawberry/pull/1100).

### New features:
- allow overriding `StrawberryField`'s
    - `get_result` - for overriding the execution of resolvers.


---
Here's the preview release card for twitter:

![](https://i.imgur.com/7aJ9QLp.png)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Nir for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```
<!-- action-check: release-file -->

patrick91 commented on pyproject.toml:
```diff
@@ -1,7 +1,7 @@
 name = "strawberry-graphql"
 packages = [ { include = "strawberry" } ]
-version = "0.126.0"
+version = "0.126.1"
```
this is done automatically when merging 😊 can you revert it?

nrbnlulu replied:
sure

patrick91 commented on strawberry/object_type.py:
```diff
@@ -45,33 +41,15 @@ def _check_field_annotations(cls: Type):
-            # >>>    raise Exception
+        if isinstance(field_, StrawberryLazyField):
+            strawberry_field = field_.evaluate(cls, field_name)
+            setattr(cls, field_name, strawberry_field.to_dataclass_field)
```
`to_dataclass_field` sounds like a function, shall we change it do be a regular function instead of a property?

nrbnlulu replied:
it is a function :upside_down_face: thats a typo

patrick91 commented on tests/fields/test_basic.py:
```diff
@@ -111,6 +108,3 @@ class Query:
     fields = Query._type_definition.fields
     assert [field.default_value for field in fields] == [3, []]
-
-    with pytest.raises(InvalidDefaultFactoryError):
-        strawberry.field(default_factory=round)
```
why is this not raising anymore?

nrbnlulu replied:
I thought that this is redundant since it is pure dataclasses logic and has nothing to do with strawberry am I wrong?

patrick91 replied:
that's a good point! I'd keep it though, mostly so that we don't accidentally remove this dataclasses logic 😊

nrbnlulu replied:
This can make it easier to generate docs later on though..

patrick91 commented on tests/fields/test_custom.py:
```diff
@@ -0,0 +1,23 @@
+
+    @strawberry.type
+    class Query:
+        hello: str = WorldField.evolve_with_resolver(resolve_hello)
```
I'm not sure `evolve_with_resolver` is easy to understand, could this follow the same api as `strawberry.field`, so:

`WorldField(resolver=resolve_hello)`?

nrbnlulu replied:
I'll have a try

patrick91 commented on strawberry/field.py:
```diff
@@ -23,129 +23,289 @@
+@dataclasses.dataclass(frozen=True)
+class StrawberryField:
+    python_name: Optional[str]
+    """
+    name to be used by python, i.e in the generated __init__
+    """
```
nit: I don't we need all these doc strings, I think they make the code less readable and don't provide a lot of value 😊

nrbnlulu replied:
This can make it easier to generate docs later on though...

patrick91 commented on strawberry/field.py:
```diff
@@ -23,129 +23,289 @@
+        directives: Sequence[object] = (),
+        dataclass_options: Optional[DataclassFieldSettings] = DEFAULT_FIELD_SETTINGS,
+        is_basic_field: bool = False,
+        # legacy compatibility, will be passed to dataclass_options:
+        default: object = UNSET,
+        default_factory: Union[Callable[[], Any], object] = UNSET,
```
why is this legacy? this is (also) part of Strawberry's api 😊

nrbnlulu replied:
I want to remove this and let users use 
`DataclassFieldSettings` instead, this will allow to modify the fields kwargs as well as provide better typing information.

also this is not related to `StrawberryField` I want to dataclass logic from strawberry logic as much as possible.

nrbnlulu replied:
What do you have in mind? Should i drop the `DataclassFieldSettings`?

patrick91 replied:
yup 😊 we can declare all the params on our function, and maybe use this in future to make it cleaner; https://peps.python.org/pep-0692/

patrick91 commented on strawberry/field.py:
```diff
@@ -23,129 +23,289 @@
-    def __call__(self, resolver: _RESOLVER_TYPE) -> "StrawberryField":
-        """Add a resolver to the field"""
+    @classmethod
+    def create(
```
why a class method and not just a constructor? :)

nrbnlulu replied:
lol I though this would allows to modify the `dataclass __init__` along the way didn't though to call `super().__init__(...)`

nrbnlulu replied:
actually I was right, you cannot assign to a frozen dataclass in __init__ see [this SO](https://stackoverflow.com/questions/57893902/how-can-i-set-an-attribute-in-a-frozen-dataclass-custom-init-method)

patrick91 commented on strawberry/field.py:
```diff
@@ -23,129 +23,289 @@
-        # Allow for StrawberryResolvers or bare functions to be provided
-        if not isinstance(resolver, StrawberryResolver):
-            resolver = StrawberryResolver(resolver)
+    @classmethod
+    def evolve_with_resolver(
+        cls,
+        instance: "StrawberryField",
+        base_resolver: StrawberryResolver,
+        origin: Union[Type, Callable, staticmethod, classmethod],
+    ) -> "StrawberryField":
```
I'm unclear why we need this and can't use standard inheritance 🤔

nrbnlulu replied:
This is not for overriding `StrawberryField` this is just me refactoring... instead of having the `StrawberryField` built up in so many places we build it only using several functions.

patrick91 replied:
mmh, what do you mean exactly?

nrbnlulu replied:
for instance field.origin is determined in `strawberry.type` IIRC it is quite cumbersome IMO.

nrbnlulu replied:
Here:
https://github.com/strawberry-graphql/strawberry/blob/1fc34fa708ebb0d21c6e1a34f86c6202fcfb4783/strawberry/types/type_resolver.py#L116

not only that it is weird that `_get_field` (that is a getter) is messing up with state, it is also unclear to me why wouldn't this logic would live inside `StrawberryField`.

patrick91 replied:
That sounds good, but I'm not sure why we need to use these methods and not use the constructor or have a single evaluate call for this 😊

nrbnlulu replied:
> use the constructor
It is impossible with frozen dataclasses.
> have a single evaluate call for this.
there actually is only one evaluate call for this in `StrawberryLazyField` :upside_down_face:  , Wym?

patrick91 replied:
do we need to make it frozen?

when is `evolve_with_resolver` called?

nrbnlulu replied:
> do we need to make it frozen?  

I like it this way, keeps things more robust.

> when is evolve_with_resolver called? 

currently [here](https://github.com/strawberry-graphql/strawberry/pull/2117/files#diff-00af9d43c9b59404ba170e72e470939f2a6dd50e2eae24d0e24ef105431e5414R454).Though  I am still trying to figure out where is the best place to call `StrawberryLazyField.evaluate`

patrick91 reviewed (commented):
Thanks so much for doing this, I've added a few comments 😊

nrbnlulu commented:
Hey, I have dropped the idea of a frozen styled class even though it makes thing more orgenized, it would be less dynamic.

Currently I am relying heavily on `__call__` of `StrawberryField`, this has several benefits:

- supports using the `StrawberryField` as a decorator.
- having only one public entry for field creation
- allows to batch fields like this:
```
class StrawberryField:
    def __init__(self, child=None)
        self.child = child

@strawberry.type
class Query:
    @UpperCaseField(AuthenticationRequiredField())
    def get_username(self, id: strawberry.ID) -> str:
        return "jkimbo"
```
this is not currently implemented but I think I'll be able to do this,

nrbnlulu commented on strawberry/types/types.py:
```diff
@@ -23,6 +29,80 @@
+
+    fields: Dict[str, StrawberryField] = {}
+
+    # TODO: What is this?
```
@patrick91 can you tell me if that is necessary? or what is that for? fields test are passing locally for me without it.

nrbnlulu commented:
Also, I have rename `_type_defenition` to `__strawberry_definition__` I hope you will forgive me not putting this in another PR, though `_type_defenition` is kinda cryptic :upside_down_face: ...

BryceBeagle commented on docs/types/exceptions.md:
```diff
@@ -200,45 +200,84 @@ MyCustomScalar2 = strawberry.scalar(
-.../venv/lib/python3.9/site-packages/graphql/type/definition.py:769: in fields
-    raise TypeError(f"{self.name} fields cannot be resolved. {error}")
-E   TypeError: Query fields cannot be resolved. Scalar `MyCustomScalar` has already been registered
+... / venv / lib / python3
```
I think the autoformatter got a bit eager here

BryceBeagle commented on tests/schema/test_lazy_types/type_a.py:
```diff
@@ -20,3 +20,5 @@ def type_b(self) -> strawberry.LazyType["TypeB", ".type_b"]:  # noqa
 
         return TypeB()
+
+    done: bool = True
```
What is this change here for? I see that it's used in `test_cyclic` but I'm wondering why. Also what drove it to be added as a resolver-less field _below_ a field with a resolver -- is that necessary?

nrbnlulu replied:
Due to how I evaluated forward refs it seems like the schema would not be generated in the same order, so I just had that test there to be sure.

BryceBeagle commented on tests/schema/test_lazy_types/test_cyclic.py:
```diff
@@ -30,5 +31,23 @@ class Query:
+    assert res.errors is None
+    assert res.data["a"]["typeB"]["typeA"]["done"]
+    expected = textwrap.dedent(expected).strip()
+    schema = print_schema(schema).splitlines()
+    for line in expected.splitlines():
+        assert line in schema
```
Why this change? Are we not guaranteeing order?

nrbnlulu replied:
See above

nrbnlulu replied:
Is order mandatory?

BryceBeagle commented on tests/fields/test_resolvers.py:
```diff
@@ -134,9 +132,11 @@ def adios(self):
-        @strawberry.field
-        def hello(self, query) -> str:
-            return "I'm a resolver"
+        @strawberry.type
+        class ShouldRaiseHere:
+            @strawberry.field
+            def hello(self, query) -> str:
+                return "I'm a resolver"
```
Not sure about this change. Is it no longer a requirement that a field is "complete" until it's added to a type?

nrbnlulu replied:
There are some stuff that cannot be determined until the origin is determined, for basic fields we must kniw the python name and the origin so fields are not evaluated at all until they are added in a type. I cant see a use case where this might cause problems, do you?

BryceBeagle commented on tests/fields/test_custom.py:
```diff
@@ -0,0 +1,9 @@
+
+
+def test_override_get_result():
+    # TODO: after patrick approves changes.
```
Let's make sure not to merge until this is resolved

nrbnlulu replied:
I'll add an assertion

BryceBeagle commented on tests/fields/test_basic.py:
```diff
@@ -109,8 +124,19 @@ class Query:
+            round_takes_arguments: int = strawberry.field(default_factory=round)
+
+
+def test_evolve_with_wrong_argument():
+    with pytest.raises(TypeError):
+
+        @strawberry.type
+        class Query:
+            the_field: int = strawberry.field(default=3)(2)
```
This test is the only usage of the term "evolve" anywhere in your change -- what does it mean? What does this test check for?

nrbnlulu replied:
When a field is called (similar to how you did this) it will be added with a resolver or something else and will be re-evaluated in __post_init__

BryceBeagle commented on strawberry/utils/typing.py:
```diff
@@ -117,3 +117,9 @@ def __dataclass_transform__(
     return lambda a: a
+
+
+def update_annotations(cls: Type, annots: dict) -> None:
```
Possible to narrow the `dict` type annotation? I believe the key will always be a `str`?

BryceBeagle commented on strawberry/exceptions.py:
```diff
@@ -246,3 +250,12 @@ def __init__(self):
+
+
+class UncallableResolverError(Exception):
+    def __init__(self, resolver: "StrawberryResolver"):
```
Quotes around `StrawberryResolver` are unnecessary because we're using `__future__` annotations

BryceBeagle commented on strawberry/annotation.py:
```diff
@@ -269,12 +275,19 @@ def _strip_lazy_type(cls, annotation: LazyType) -> type:
-def _is_object_type(type_: Any) -> bool:
-    # isinstance(type_, StrawberryObjectType)  # noqa: E800
-    return hasattr(type_, "_type_definition")
+def _get_strawberry_definition(type_: Any) -> Optional[StrawberryDefinition]:
+    origin = type_
+    # generics store their class in __origin__
+    if origin_ := getattr(type_, "__origin__", False):
+        origin = origin_
+    res = getattr(origin, "__strawberry_definition__", None)
+    if isinstance(res, StrawberryDefinition):
+        return res
+    else:
+        return None
```
You implemented this function (which I think is great!), but there are a lot of places where we're still manually grabbing the `__strawberry_definition__` off the types. I think it should be:
1. be moved out of `annotations.py`
2. used in more places

nrbnlulu replied:
Where do you think it should live?

BryceBeagle commented on strawberry/types/types.py:
```diff
@@ -23,26 +33,257 @@
 
 @dataclasses.dataclass(eq=False)
-class TypeDefinition(StrawberryType):
+class StrawberryDefinition(StrawberryType):
```
1. @patrick91, what do you think of this renaming. I wonder if calling this `StrawberryDefinition` is too broad and this should be `StrawberryObjectDefinition` or `StrawberryTypeDefinition`
2. @nrbnlulu, I hate to ask this because it's going to turn into a fair bit of work for you but can you please do this renaming in a separate diff? This turns a large diff into an absolutely massive diff making this fields change really hard to review.

    This renaming of `TypeDefinition` -> `StrawberryDefinition` is also a somewhat breaking change -- we want to separate it from the fields change to allow easier version upgrading by our users

BryceBeagle replied:
> Also, I have rename `_type_defenition` to `__strawberry_definition__` I hope you will forgive me not putting this in another PR, though `_type_defenition` is kinda cryptic :upside_down_face: ...

Just saw this comment :grimacing:. I think splitting it out is pretty important, unfortunately

patrick91 replied:
doing it in another PR sounds best :) this will break strawberry django and other libs I think :)

for the name I'd go with StrawberryTypeDefinition for object types, StrawberryInputTypeDefinition for input types and StrawberryInterfaceDefinition for interfaces. But that makes me think of how we have StrawberryScalar, which is not too consistent with these names, maybe we can rename it to StrawberryScalarDefinition (something for another PR )

nrbnlulu replied:
O.K. although we can have a property with a warning that will return a pointer to the new name...

BryceBeagle reviewed (changes requested):
I've done an introductory review of the change. I didn't have much time to review tonight so I did everything _except_ look at the important, logic-changing part of the change. That should also be easier if the `TypeDefinition` renaming is pulled out into a separate change

nrbnlulu commented:
@patrick.py#0235 @ignormies#0001 I have done major refactoring here, there was some things that I just couldn't do with the previous way they were.
I hope that you will find this helpful as I dedicated most of my time recently just for this PR.

Some notes about major changes:
 - I didn't touched the public API
 - I introduced new `validate` abstract method to `StrawberryType` it is used to determine generic unions
 - I have removed `copy_with` and `type_params` because I couldn't get it to work with the new way `StrawberryField` is determined, also it was very complicated IMO. The alternative now is a new `TemplateTypeDefinition` class, if a class is inheriting a `Generic` it would be instanciated as a template by `@strawberry.type`. This class can generate a whole new "`StrawberryObject`" based on `__args__` with just two functions and it also has a cache map based on `__args__` tuple.

nrbnlulu commented on strawberry/field.py:
```diff
@@ -9,292 +8,305 @@
+        if not isinstance(self.base_resolver, StrawberryResolver):
+            if resolver := resolveable(self.base_resolver):
+                self.base_resolver = StrawberryResolver(resolver)
+            #     TODO: Tests this fails
```
forgot this todo

nrbnlulu replied:
done.

nrbnlulu commented on strawberry/field.py:
```diff
@@ -394,24 +443,29 @@ def field(
+        return field_
+
+    # called like @strawberry.field(description="ABC") or not as decorator
+    # further on the decorator would call StrawberryLazyField.__call__(decorated)
```
no more `StrawberryLazyField` :smile:

nrbnlulu commented on strawberry/type.py:
```diff
@@ -142,3 +126,17 @@ def __eq__(self, other) -> bool:
         return super().__eq__(other)
+
+
+class StrawberryGqlWrapper(StrawberryType):
```
I think it is not used anymore, to delete.

nrbnlulu commented on strawberry/types/types.py:
```diff
@@ -23,6 +38,176 @@
+    return interfaces
+
+
+def _get_fields(cls: Type) -> List["StrawberryField"]:
```
maybe the same as the new _eval_fields.

nrbnlulu commented on tests/fields/test_arguments.py:
```diff
@@ -145,6 +145,8 @@ def name_resolver(
     class NameQuery:
         name: str = strawberry.field(resolver=name_resolver)
 
+    assert hasattr(NameQuery, "_type_definition")
```
replace with `get_type_definition`

nrbnlulu commented:
I tried to fix some pydantic tests though I don't really understand much there, it seems like they were relying on `copy_with` and it is needed to be migrated to `TemplateTypeDefinition.generate()`

nrbnlulu commented on strawberry/schema/name_converter.py:
```diff
@@ -94,27 +92,42 @@ def from_union(self, union: StrawberryUnion) -> str:
+        generated types needed to be renamed, to avoid naming conflicts.
+        """
+        template = generic_type.concrete_of
+        assert template
```
Just assert  `generic_type.concrete_of`

nrbnlulu commented on tests/objects/generics/test_names.py:
```diff
@@ -41,8 +41,8 @@ class TypeB:
         ([CustomInt], "CustomIntExample"),
-        ([TypeA, TypeB], "TypeATypeBExample"),
-        ([TypeA, LazyType["TypeB", "test_names"]], "TypeATypeBExample"),  # type: ignore
+        ([TypeB], "TypeBExample"),
```
I am uncertain what were these params, this is invalid generation, this class takes only one generic we should probably provide an error if someone tries this.

nrbnlulu commented on tests/test_repr.py:
```diff
@@ -12,8 +12,10 @@ class MyType:
         id: strawberry.ID
 
+    instance = MyType("a", 1, True, 3.2, "123")
+    instance.__repr__()
```
revert this.

patrick91 commented on RELEASE.md:
```diff
@@ -0,0 +1,7 @@
+Release type: patch
```
```suggestion
Release type: minor
```
I think this deserves a minor release 😊

patrick91 commented on strawberry/custom_scalar.py:
```diff
@@ -31,42 +33,80 @@ def identity(x):
-        self, type_var_map: Mapping[TypeVar, Union[StrawberryType, type]]
-    ) -> Union[StrawberryType, type]:
-        return super().copy_with(type_var_map)
+    def to_graphql_core(self):
```
can we keep this in the schema converter?

nrbnlulu replied:
I'll have a try.

patrick91 commented on strawberry/arguments.py:
```diff
@@ -152,9 +152,9 @@ def convert_argument(
-    if hasattr(type_, "_type_definition"):  # TODO: Replace with StrawberryInputObject
-        type_definition: TypeDefinition = type_._type_definition  # type: ignore
-
+    if type_definition := get_type_definition(
```
we can't use python 3.8 syntax just yet :)

also can we make get_type_definition "private"? I don't we should expose it just yet to users 😊

nrbnlulu replied:
Sure

patrick91 commented on strawberry/custom_scalar.py:
```diff
@@ -31,42 +33,80 @@ def identity(x):
+                return True
+        return False
 
-class ScalarWrapper:
```
why did you remove this? :)

nrbnlulu replied:
check my comment: https://github.com/strawberry-graphql/strawberry/pull/2117#issuecomment-1231474766

> I removed ScallarWrapper as it made issued for me with the new validate method of StrawberryType, every scalar now is just a ScalarDefinition, I would consider renaming this to StrawberryScalar.

patrick91 replied:
is it needed for this PR?

nrbnlulu replied:
Yes. it is how generic unions would be resolved here.

https://github.com/strawberry-graphql/strawberry/pull/2117/files#diff-111023ca6e4b982dfea8bea22a5eafe990b8639010fc30dbd7e515a96fc16fb2R120

patrick91 commented on strawberry/field.py:
```diff
@@ -9,292 +8,305 @@
-
-    def __init__(
+@dataclasses.dataclass()
+class StrawberryField(StrawberryType):
```
I'm not sure this should extend StrawberryType /cc @BryceBeagle

nrbnlulu replied:
This is also for `validate()` see https://github.com/strawberry-graphql/strawberry/pull/2117#issuecomment-1231474766

patrick91 reviewed (commented):
thanks for putting so much work into this, I'm having troubles finding enough time to review this unfortunately, is there a way we can break it into multiple PRs? I've seen some changes that don't seem strictly related to the custom field work, it might be best to keep them separated

patrick91 commented:
@nrbnlulu ok, I'll reach out to you in the next few days 😊 thanks for making time to go through the code

nrbnlulu commented on tests/experimental/pydantic/test_conversion.py:
```diff
@@ -844,52 +839,6 @@ class UserType:
     assert user.passwords == ["hunter2"]
 
 
-def test_sort_creation_fields():
```
I dropped the `DataclassCreationField` it is no longer needed with the new `StrawberryObejct` field creation hook.

nrbnlulu replied:
@patrick91 do you approve this? also when #1187 is merged this is probably useless.

nrbnlulu commented on tests/experimental/pydantic/test_conversion.py:
```diff
@@ -946,32 +895,6 @@ class User(BaseModel):
             password: Optional[str] = Field(default=None, default_factory=lambda: None)
 
 
-def test_can_convert_pydantic_type_to_strawberry_with_additional_field_resolvers():
```
This test is not a valid strawberry behavior. strawberry doesn't create dataclasses fields for fields with resolvers. so I removed this for now. let me know what you think.

nrbnlulu replied:
@patrick91 what do you think?

nrbnlulu commented on strawberry/experimental/pydantic/conversion.py:
```diff
@@ -72,12 +72,9 @@ def convert_pydantic_model_to_strawberry_class(cls, *, model_instance=None, extr
 
-        # only convert and add fields to kwargs if they are present in the `__init__`
-        # method of the class
-        if field.init:
```
test are not affected by this but this should probably be reverted

nrbnlulu commented on tests/schema/test_generics.py:
```diff
@@ -135,6 +136,21 @@ class Query:
 
+    from strawberry.utils.typing import is_generic_subclass
+
+    class StrawberryGenericsGenerator:
```
(: I think thats a blooper

nrbnlulu commented on tests/schema/test_generics.py:
```diff
@@ -492,6 +536,10 @@ class Edge(Generic[T]):
         cursor: strawberry.ID
         node: T
 
+    assert Edge[int]._type_definition.fields[1].type is int
```
can be removed.

nrbnlulu commented:
@patrick91 Any updates on this one?

patrick91 commented:
@nrbnlulu not yet sorry, I'm travelling for conferences so I don't have a lot of focused time 😊
